### ISSUE TITLE
compatibility with Kibana

### ISSUE THREAD

**Issue by githubmanticore** [nck-2/test-rep#244]
Let's make Manticore Search compatible with Kibana: 
  * the basic test can be found [here](https://gitlab.com/manticoresearch/docker/tree/master/logs), feel free to make any changes you need 
  * let me know if something is not clear in the set of dockers (e.g. in what order they should be run etc.) 
  * the task is to make it possible to replace Elasticsearch with Manticore search on the port Kibana uses to access Elasticsearch so that basic functionality works 
  * we don't need to make Manticore compatible with some things too specific to Elasticsearch. I'm not an expert or at least an advanced user of Kibana though and can't say now what those things are. It requires a good research. 
  * perhaps it makes sense to start with some python script as an adapter between Kibana and Manticore Search just to move faster and only when the logic is clear and works properly port it to C++. But it's up to you.

**Comment by githubmanticore** [comment]
➤ Stan commented:
 

 
I've just pushed initial Kibana compatibility mode at kibana_wip branch. To make daemon work as Elastic backend for Kibana someone should: 
- start daemon at configless mode with HTTP interface 
- issue command `mysql -h 127.0.0.1 -P 8306 -e "set global log_management = 1"` to switch HTTP interface to be Elastic compatible 
- after that start Kibana with `elasticsearch.hosts` as Manticore daemon host and port

**Comment by githubmanticore** [comment]
➤ Stan commented:
 

 
tables that will be used for Kibana as data source should be created with `index_field_lengths='1'` option, like 
``` 
CREATE TABLE weblog1 ( 
idd integer, 
@timestamp integer, 
response_time integer, 
bytes integer, 
code integer, 
httpversion string attribute indexed, 
agent string attribute indexed, 
referer string attribute indexed, 
verb string attribute indexed, 
message string attribute indexed, 
request string attribute indexed, 
@version string attribute indexed, 
upstrm string attribute indexed, 
ip string attribute indexed, 
title text indexed 
) index_field_lengths='1'; 
``` 
 
after Kibana starts someone should create index pattern, wildcards works for index pattern however all matched indexes at Manticore should have the same schema.

**Comment by githubmanticore** [comment]
➤ Stan commented:
 

 
currently at searchd daemon all uservars statement and select list expression got broken and returns parser error. 
 
It worth to check the Kibana compatibility mode and make decision to keep developing it or not. 
In case of further development someone should collect typical work scenarios for Kibana and what are broken now and should be fixed then 
- fix broken cases - create more transforms for incoming requests 
- fix unoptimal reply processing - now there are multiple `string -> json -> string` transforms of result-set these should be eliminated 
- fix broken uservars 
- check `nljson` library usage and decide to keep +1 JSON processing library or get rid of it and rewrite JSON transforms with current JSON processing libraries that we already have 
- after all typical scenarios will known and work - need to eliminate table properties or whole tables from `http/log_management.h` or generate these tables at runtime. 
- create Selenium automation testing for typical cases of Kibana at GitLab CI pipeline as currently fixing one part of code might break another part that got discovered much later

**Comment by githubmanticore** [comment]
➤ Stan commented:
 

 
I used docker-compose for testing and it seems that mitmproxy took place in communication with searchd that is why Kibana does not start as `docker run -e ELASTICSEARCH_HOSTS=http://88.99.251.227:18380 -p 25601:5601 docker.elastic.co/kibana/kibana:7.4.2` with only daemon. 
 
I've just pushed fixes at kibana_wip branch: 
 
- handle of HEAD requests. As `mitm` switch HEAD to GET and that was not handled at daemon. And also daemon reply with body to HEAD requests and `mitm` cut the body on send reply to Kibana. 
 
I added HEAD handling for some requests to not reply with body but only with actual Content-Length option. 
 
- fixed long strings at log_management.h and JSON object update these prevent Kibana starting from scratch with daemon
